pep-9999.rst

```diff
@@ -0,0 +1,257 @@
+
+
+Specification
+=============
+
+The specification consists of three parts: the configuration definition, package discovery and configuration file discovery.
+
+
+Configuration definition
+------------------------
+
+The configuration is split into two files:
+
+* ``config.toml`` for PyPI repository configuration
+* ``auth.toml`` for credentials
+
+
+The config.toml file
+''''''''''''''''''''
+
+Package index configuration is defined in the ``config.toml`` file::
+
+
+    [repositories]
+
+    [repostiories.pypi]
+    url = "https://pypi.org/simple/"
+
+    [repostiories.myrepo1]
+    url = "http://myserver2.org/simple"
+
+    [authentication]
+
+    [authentication.pypi]
+    username = "__token__"
+    password = "pypi-AgEIcHlwaS5vcmcCJDJjMmI5NDJmLWIzNzQtNDc3ZC04ZmRkLTM4NDEzY2M1ZTU5MAACJXsicGVybWlzc2lvbnMiOiAidXNlciIsICJ2ZXJzaW9uIjogMX0AAAYgV91r8iEUUz7Q5xbsfi_KGiO3wrJR2_IOOpnFyLo62bk"
+
+    [authentication.myrepo1]
+    username = "foo"
+    password = "YmFyCg=="
+
+
+A repository has the following attributes::
+
+    ==================  ===========  =============================================
+    Attribute           Type         Description
+    ==================  ===========  =============================================
+    final               bool         Make it impossible to override/remove the repository configuration (default: false)
+    priority            int          Order of priority, sorted descending (default: 0)
+    proxy_url           str          The URL to the proxy server, can contain a port
+    shadow              bool         Use package from this index, even if it was found in a previous index (default: false)
+    url                 str          URL to repository, can contain a port
+    ==================  ===========  =============================================
+
+
+A repository's credentials has the following attributes::
+
+    ==================  ===========  ===============================================
+    Attribute           Type         Description
+    ==================  ===========  ===============================================
+    cert                str          Path to certificate authority bundle (*.pem)
+    cert-client         str          Path to client certificate (*.pem)
+    enabled             bool         Whether the repository should be used or not (default: true)
+    password            str          The password for the repository
+    proxy_password      str          The password for the proxy server
+    proxy_username      str          The username for the proxy server
+    username            str          The username for the repository
+    ==================  ===========  ===============================================
+
+
+Tools preferring to look up credentials from other locations, such as keyring, are advised to use the full repository URL as the key, rather than the index name or hostname.
```

Comment:
Should expand here as to why this is important. Since package repos like Azure DevOps and GitLab Artifacts have per repo index urls.

Further the PEP can allow for "hostname" matching to be fallback when a more specific url is not available. Since otherwise, we end up source vs publish end point issues.

Reply:
As I don't have enought insight here, I would love for you to help writing this. Would you be up for that?